jupyterhub/tests/test_db.py

```diff
@@ -1,45 +1,70 @@
 from glob import glob
 import os
-import shutil
+from subprocess import check_call
+import sys
+import tempfile
 
 import pytest
 from pytest import raises
 from traitlets.config import Config
 
-from ..dbutil import upgrade
 from ..app import NewToken, UpgradeDB, JupyterHub
 
 
-here = os.path.dirname(__file__)
-old_db = os.path.join(here, 'old-jupyterhub.sqlite')
+here = os.path.abspath(os.path.dirname(__file__))
+populate_db = os.path.join(here, 'populate_db.py')
 
-def generate_old_db(path):
-    db_path = os.path.join(path, "jupyterhub.sqlite")
-    print(old_db, db_path)
-    shutil.copy(old_db, db_path)
-    return 'sqlite:///%s' % db_path
 
-def test_upgrade(tmpdir):
-    print(tmpdir)
-    db_url = generate_old_db(str(tmpdir))
-    upgrade(db_url)
+def generate_old_db(env_dir, hub_version, db_url):
+    """Generate an old jupyterhub database
 
+    Installs a particular jupyterhub version in a virtualenv
+    and runs populate_db.py to populate a database
+    """
+    env_pip = os.path.join(env_dir, 'bin', 'pip')
+    env_py = os.path.join(env_dir, 'bin', 'python')
+    check_call([sys.executable, '-m', 'virtualenv', env_dir])
```

Comment:
Can this be 'venv' instead of 'virtualenv'? Might alleviate the need to install virtualenv in requirements

Reply:
Test failed locally when running using virtualenv. Test passed when using venv. I'm running all in python36 environment.

Reply:
It failed before with venv, I can't recall why. Maybe missing `ensurepip`? `virtualenv` has been added to the test requirements. Have you re-run `pip install -r dev-requirements.txt` from this PR?